tests/hil/scripts/pytest-hil/board.py

```diff
@@ -0,0 +1,63 @@
+from abc import ABC, abstractmethod
+import serial
+from time import time
+import re
+import yaml
+
+class Board(ABC):
+    def __init__(self, port, baud, credentials_file):
+        self.serial_device = serial.Serial(port, baud, timeout=1)
+
+        # Read credentials
+        with open(credentials_file, 'r') as f:
+            credentials = yaml.safe_load(f)
+            settings = credentials['settings']
+
+        # Set PSK credentials
+        self.set_golioth_psk_credentials(settings['golioth/psk-id'], settings['golioth/psk'])
+
+        # Set WiFi credentials
+        if self.USES_WIFI:
+            self.set_wifi_credentials(settings['wifi/ssid'], settings['wifi/psk'])
+
+    def wait_for_regex_in_line(self, regex, timeout_s=20, log=True):
+        start_time = time()
+        while True:
+            line = self.serial_device.readline().decode('utf-8', errors='replace').replace("\r\n", "")
+            if line != "" and log:
+                print(line)
+            if time() - start_time > timeout_s:
+                raise RuntimeError('Timeout')
+            regex_search = re.search(regex, line)
+            if regex_search:
+                return regex_search
+
+    def send_cmd(self, cmd, wait_str=None):
+        if wait_str == None:
+            wait_str = self.PROMPT
+        self.serial_device.write('\r\n'.encode())
+        self.wait_for_regex_in_line(self.PROMPT)
+        self.serial_device.write('{}\r\n'.format(cmd).encode())
+        self.wait_for_regex_in_line(wait_str)
+
+    @property
+    @abstractmethod
+    def PROMPT(self):
+        pass
+
+    @property
+    @abstractmethod
+    def USES_WIFI(self):
+        pass
```

Comment:
Do we have some plan or coding guidelines when to use `UPPER_CASE` methods? I don't want to be too strict with tooling that is not core of our product, but I didn't see such naming convention before in Python projects, I think.

Personally I found `flake8` and `pylint` to be used most commonly, but this is a very subjective point of view. Overall, most Python linters or code checker tools are standardized around PEP 8: https://peps.python.org/pep-0008/ anyway.

Reply:
The idea here is that `USES_WIFI()` should be a class constant defined by the subclasses. Python doesn't have constants, but it does have [class properties](https://docs.python.org/3/library/functions.html?highlight=property#property). Because we only define a getter for this property (and no setter), and because it returns a constant value (subclasses return either `True` or `False), it acts as constant. Because it has the property decorator, you can access it with `board.USES_WIFI` (no parenthesis). So even though it is defined as a method,  it is used as a constant property, which is why I made it upper case. I don't know of any guidelines around this kind of thing, but to me it makes the most sense and indicates the intention in the clearest way.

Reply:
And how about using class attribute, like that:
```
class Board:
    USES_WIFI = False


class WiFiBoard(Board):
    USES_WIFI = True


class CellularBoard(Board):
    pass


if __name__ == '__main__':
    board = Board()
    wifi = WiFiBoard()
    cellular = CellularBoard()

    print('board', board.USES_WIFI)
    print('wifi', wifi.USES_WIFI)
    print('cellular', cellular.USES_WIFI)
```
which resolves to:
```
board False
wifi True
cellular False
```

Reply:
Class attributes are mutable:

```
class Board:
    USES_WIFI = False


class WiFiBoard(Board):
    USES_WIFI = True


class CellularBoard(Board):
    pass


if __name__ == '__main__':
    board = Board()
    wifi = WiFiBoard()
    cellular = CellularBoard()

    print('board', board.USES_WIFI)
    print('wifi', wifi.USES_WIFI)
    wifi.USES_WIFI = False
    print('wifi', wifi.USES_WIFI)
    print('cellular', cellular.USES_WIFI)
```
gives:
```
board False
wifi True
wifi False
cellular False
```

Using properties makes it immutable:

```
class Board:
    @property
    def USES_WIFI(self):
        return False


class WiFiBoard(Board):

    @property
    def USES_WIFI(self):
        return True


class CellularBoard(Board):
    pass


if __name__ == '__main__':
    board = Board()
    wifi = WiFiBoard()
    cellular = CellularBoard()

    print('board', board.USES_WIFI)
    print('wifi', wifi.USES_WIFI)
    wifi.USES_WIFI = False
    print('wifi', wifi.USES_WIFI)
    print('cellular', cellular.USES_WIFI)
```

```
board False
wifi True
Traceback (most recent call last):
  File "/Users/samfriedman/golioth/test_cp_sdk_zephyr/modules/lib/golioth-firmware-sdk/test.py", line 25, in <module>
    wifi.USES_WIFI = False
    ^^^^^^^^^^^^^^
AttributeError: property 'USES_WIFI' of 'WiFiBoard' object has no setter
```

Reply:
I am not fond of using `@property` as a workaround for immutability (that is the nature of Python, there is even no concept of private or protected methods, since those are just "protected" by convention and nothing else). But But I don't won't to block just because of that reason.

Reply:
I think even more than just being immutable, it provides a cleaner getter/setter interface for variables, which is generally preferred over accessing object variables directly in OOP. In the future, if the logic for whether or not a board uses WiFi is more complex (e.g. if there is a SOC with WiFi and Cell support), the functions can be extended.

Reply:
Actually there are plenty of libraries, or even all libraries, that use some kind of attributes. Either on the class or instance level. The same reasoning for immutability and getter/setter interface could be used for them. I can't imagine that converting all of the code to use `@property` in order to achieve immutability would be welcome. For the same reason I don't think it is a good idea to introduce such pattern in this module.

Besides, this property is written in UPPER_CASE, which means that it should not be overridden, just by convention.

> In the future, if the logic for whether or not a board uses WiFi is more complex (e.g. if there is a SOC with WiFi and Cell support), the functions can be extended.

In that case we should make that property instance specific and use snake_case name pattern.

Right now implementation says one thing, while name of the method suggests something else. Reliance on convention is quite important in Python, at least that is my feeling, so I am all for following it instead of inventing our own.

Reply:
There are also plenty of libraries using the @property decorator - this is a built-in feature of Python, so saying that it is uncommon or novel to use doesn't make much sense.

There are many ways of accomplishing the same goal - even in Python - and it's ok if the code I write doesn't look exactly like the code that you write and vice versa.